### PR TITLE
switch CI to GitHub Actions

### PR DESCRIPTION
**Pull request by benoit-pierre**
# Summary

With [Travis migration from `.org` to `.com` and the associated changes regarding their Open Source offering](https://travis-ci.community/t/org-com-migration-unexpectedly-comes-with-a-plan-change-for-oss-what-exactly-is-the-new-deal/10567/30), I decided to check out Github Actions as a replacement for all our CI needs.

# Pros & Cons

## Pros:
- 2 less services to wrestle with
- the macOS build will be in line with others: done on the result of merging to master (instead of the PR branch on Circle CI), so closing/reopening a PR will suffice to rerun all the CI checks updated with the latest changes from master
- no weird access restrictions for the macOS build like with Circle CI
- when making a new release, creating the draft on Github can be done in one step, and only if all the checks were successful: no more having to manually merge multiple drafts
- artifacts for all 3 platforms in one place (including the Linux AppImage)

## Cons:
- no more SSH support for debugging the macOS build (really the only feature I will miss from Circle CI)
- the macOS jobs can be really slow (according to [this issue](https://github.com/actions/virtual-environments/issues/2707), things should improve in the next couple of months)
- I find Github Actions lacking a little with respect to helping avoid duplication when defining jobs: the main YAML configuration does not support anchors / aliases, their composite actions don't support re-using other actions, ...
- the Github Actions UI can be a little flaky (but the workflow runs will execute correctly)

Note: this PR depends on (and contains): #1195, #1196, #1197, #1201, and #1202.

# Github Actions Implementation

To reduce code duplication, I ended up using [Jinja](https://palletsprojects.com/p/jinja/) to generate the finally CI workflow configuration (same templating engine as used by Towncrier).

A run consist of:
- 3 "platform tests" jobs: Linux, macOS, and Windows
- 3 "platform build" jobs, again: Linux, macOS and Windows, dependent on their respective "platform tests" job
- 1 "packaging" jobs that run a number of packaging related checks
- 3 "Python tests": for checking support for older/newer versions of Python (other than the one currently used for our distributions)
- 1 final "release" job that is only run when making a release and depends on all the other jobs: this job will release both on PyPI and prepare a draft on Github (including assets and release notes)

All those jobs use caching to reduce times.

The macOS jobs don't use one of the Python interpreters provided by Github Actions, but instead install Python from an official installer: to support older macOS releases. On that subject, I set `MACOSX_DEPLOYMENT_TARGET` to `10.9`, which from what I gathered is used by other Python projects to build wheels that support Mavericks and later, but as always, I have no way to test that it does work.

Note: the "Python tests" are run on Linux, the rational is that on Windows/macOS, we assume people are going to use our distributions, but on Linux, Plover may be packaged by the distributions and use a different version of Python.

# Other changes

I made a number of other changes as part of supporting this:
* the requirements have been split: there are now a bunch of small files (split by use) with unpinned requirements, and one big file (`reqs/constraints.txt`) will all requirements pinned. This make it easier to install only the requirements needed for particular task, and reduce duplications.
* I moved the log related dependencies to a dedicated `log` extra
* I reworked how the Windows distribution is built: a bash script is now used (with Linux support). The old `window/helper.py` code has been removed.
* I reworked how the macOS distribution is built: same technique as in #1170, the old technique was not working

### CODE REVIEW

**Comment by benoit-pierre** [comment]
I forgot to mention that we need to configure a `release` environment with a `PYPI_TOKEN` secret set to a valid PyPI token for the release job.

Also, for an example run, see here: https://github.com/benoit-pierre/plover/actions/runs/709104578

**Comment by morinted** [comment]
This is amazing!

Is it possible to copy relevant parts of the documentation from the PR description into a Markdown file in the `ci` folder, with links to relevant files? It would help with maintaining this for anyone not familiar with the GitHub actions and workflow steps (*raises hand*)

**Comment by benoit-pierre** [comment]
I added some documentation in `doc/ci.md`.

File: doc/ci.md [hunk 1]
```diff
@@ -0,0 +1,85 @@
+# Continuous Integration
+
+The project use Github Actions (GA), the workflow configuration lives in
```
Comment (morinted) [inline doc/ci.md:3]: nit grammar

```suggestion
The project uses Github Actions (GA), the workflow configuration lives in
```

File: doc/ci.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+- `.github/workflows/ci/workflow_template.yml` is the
+  [Jinja](https://palletsprojects.com/p/jinja/) template.
+
+- `.github/workflows/ci/workflow_context.yml` contains the tests/build jobs
+  definitions.
+
+- `.github/workflows/ci/workflow_generate.py` is used to generate the workflow
+  configuration: just execute `./.github/workflows/ci/workflow_generate.py` to
+  update the workflow configuration after updating one of the files above (the
+  script will check the output is valid YAML, and that anchors/aliases are not
```
Comment (morinted) [inline doc/ci.md:17]:
```suggestion
  script will check whether the output is valid YAML, and that anchors/aliases are not
```

File: doc/ci.md [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+  used).
+
+- `.github/workflows/ci/helpers.sh` contains the bash functions used by some
+  of the steps (e.g.: setting up the Python environment, running the tests,
```
Comment (morinted) [inline doc/ci.md:21]:
```suggestion
  of the steps (e.g. setting up the Python environment, running the tests,
```

File: doc/ci.md [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  etc...). Note: this file reuse some of the helpers provided by
+  `plover_build_utils/functions.sh`.
+
+The current workflow consists of:
+
+- 3 "platform tests" jobs: Linux, macOS, and Windows
+- 3 "platform build" jobs, again: Linux, macOS and Windows, dependent on their
+    respective "platform tests" job (so if the `Test (macOS)` job fails, the
+    `Build (macOS)` job is skipped).
+- 1 "packaging" jobs that run a number of packaging related checks (see
```
Comment (morinted) [inline doc/ci.md:31]:
```suggestion
- 1 "packaging" job that run a number of packaging related checks (see
```

File: doc/ci.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  `packaging_checks` in `plover_build_utils/functions.sh`
+- 3 "Python tests" jobs: for checking support for older/newer versions of Python
+  (other than the version currently used for the distributions)
+- 1 final, optional, "release" job
+
+## Tests / Build jobs
+
+On Linux / Windows, the standard GA action `actions/setup-python` is used
+to setup Python: so for example configuring a job to use 3.7 will
```
Comment (morinted) [inline doc/ci.md:40]:
```suggestion
to setup Python: so, for example, configuring a job to use 3.7 will
```

File: doc/ci.md [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+automatically setup up the most recent 3.7.x version available on the
+runner.
+
+On macOS, to support older releases, Python will be setup from an official
+installer (see `osx/deps.sh` for the exact version being used). So the
```
Comment (morinted) [inline doc/ci.md:45]:
```suggestion
installer (see `osx/deps.sh` for the exact version being used). The
```

File: doc/ci.md [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+version declared in `workflow_context.yml` must match, or an error will
+be raised during the job execution (if for example the job is declared
+to use `3.7`, but the dependency in `osx/deps.sh` uses `3.6.8`).
+
+Caching is used to speedup the jobs. The cache is keyed with:
```
Comment (morinted) [inline doc/ci.md:50]:
```suggestion
Caching is used to speed up the jobs. The cache is keyed with:
```

File: doc/ci.md [hunk 8]
```diff
@@ -0,0 +1,85 @@
...
+- the `epoch` defined `workflow_context`: increasing it can be used to
+  force clear all the caches for all the jobs
+- the name of the job
+- the full Python version setup for the job (so including the patch number)
+- a hash of part of the requirements (`reqs/constraints.txt` + the relevant
+  `reqs` files for the job in question), and additional files declaring
+  extra dependencies for some jobs (e.g. `osx/deps.sh` on macOS)
+
+If key changes, the cache is cleared/reset, and the Python environment will
```
Comment (morinted) [inline doc/ci.md:59]:
```suggestion
If the key changes, the cache is cleared/reset, and the Python environment will
```

File: doc/ci.md [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+be recreated, wheel and extra dependencies re-downloaded, etc...
+
+## Packaging job
+
+This job will run a number of packaging related checks. See
```
Comment (morinted) [inline doc/ci.md:64]:
```suggestion
This job will run a number of packaging-related checks. See
```

File: doc/ci.md [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+`packaging_checks` in `functions.sh` for the details.
+
+The resulting source distribution and wheel will also be added
+to the artifacts when a release is being created.
+
+## Release job
+
+The final job, only run on release (tag), and if all the other jobs
+completed successfully.
```
Comment (morinted) [inline doc/ci.md:73]:
```suggestion
The final job, it only runs on release (tag), and if all the other jobs
complete successfully.
```

File: doc/ci.md [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+this to work, a valid [PyPI token](https://pypi.org/help/#apitoken)
+must be configured: the `PYPI_TOKEN` secret of the `release`
+[environment](https://docs.github.com/en/actions/reference/environments)
+will be used. Additionally, the optional `PYPI_URL` secret can be set to
+use another PyPI compatible index (e.g. Test PyPI).
+
+If the previous step was successful, a new release draft will be created
+on GitHub. All the artifacts will be included as assets. The release notes
+are automatically generated from the last release section in `NEWS.md` and
+the template in `.github/RELEASE_DRAFT_TEMPLATE.md`.
```
Comment (morinted) [inline doc/ci.md:85]: Do we want to add a `## Limitations` section that's just a bulleted list with

```
- The artifact upload action [always wraps artifacts in a zip](https://github.com/actions/upload-artifact/issues/39), even if they are a single file such as an exe or a dmg.
```

**Review by morinted — CHANGES_REQUESTED** [review]
Very nit grammar + suggestion to add a limitations section with the zipping of artifacts. Excellent docs! I understand way better what's going on now.

File: doc/ci.md (hunk 11 above)
Comment (benoit-pierre) [reply doc/ci.md:85]: OK.

**Review by morinted — APPROVED** [review]

**Comment by benoit-pierre** [comment]
[GitHub Actions is experiencing degraded performance](https://www.githubstatus.com/)... Let's wait for the issue to be resolved.

**Comment by benoit-pierre** [comment]
And... it's back.

**Pull request merged**